### ISSUE TITLE
Dynamic Xray plugin

### ISSUE THREAD

**Issue by smartscripts-nl** [koreader/koreader#12398]
**Does your feature request involve difficulty completing a task?**
KOReader doesn't have a Xray feature like Kindle has.

**Describe the solution you'd like**
- This plugin should basically gives users the option to dynamically add and manage xray items. To be used for study or reference.
- These items can be persons or things, ideas, lemmas etc., to which the user adds (written or copied) descriptions.
- Persons or things can be differentiated by icons (e.g. a user and a lightbulb icon).
- Important persons or things can be marked by dark icons.
- The xray-items can be attached to an ebook or an ebook series. In the latter case, upon opening the next book in a series the xray items will be immediately available again.
- The plugin acts as a kind of plugin to the ReaderDictionary plugin: upon clicking on a word or name in the reader, if a xray item is found that case sensitive contains (part of) the selected text, the xray item viewer will be shown instead of the Dictionary results dialog.
- When multiple xray items are found that have a bearing on the selected text, a ButtonDialog will be shown with these items, from which the user can choose the most relevant. Also this dialog has a plus-button, with which to define a new xray item from the text selection.
- When the user views a single xray item in the xray item viewer, related items are shown as buttons at the bottom of that dialog.
- The default ReaderHighlight popup for selected texts has a plus-button to define a new xray item from the text selection.
- The list of xray items is filterable by text, important items, persons, things.
- The list has a button to switch between (items attached to) series or book mode.
- Here are some examples:
![readerhighlight-selected-text-popup-with-add-xray-item-button](https://github.com/user-attachments/assets/47062ed5-acd6-4b5c-94ea-4daf13aeb5fa)
![multiple-xray-items-found-via-readerhighlight](https://github.com/user-attachments/assets/d253280c-85b0-48df-977f-2d7817110525)
![view-xray-item-found-via-readerhighlight-selection](https://github.com/user-attachments/assets/d5fbe780-f17f-4bac-8a63-5c16aa655c11)
![add-xray-item](https://github.com/user-attachments/assets/08d3f002-ad84-45ad-83e5-c1c62e6675af)
![xrays-items-list](https://github.com/user-attachments/assets/b0023d6b-3aaa-4c75-8126-3233850c1438)
![filter-xray-items](https://github.com/user-attachments/assets/05b3b8da-17f5-42bd-8222-59ef98ef1594)
![xray-items-list-filtered](https://github.com/user-attachments/assets/eab3edc5-4e3e-44c7-8532-f85c506c06f3)
![toggle-ebook-or-series-mode](https://github.com/user-attachments/assets/c075d684-1540-4bba-afc9-f0a039d95246)

**Additional context**
I myself don't need this, because I've got it working. But maybe there are other persons who would like this feature. I use an extensive extensions system, written by me, that doesn't adhere to the KOReader code writing style.  So that's no solution to the problem. I won't share that system, because it is personal to me, and also I have no wish or time to become a KOReader developer. So I post this concept, that the KOReader developers can develop this in correct KOReader code, should they wish so.

**Comment by Frenzie** [comment]
https://github.com/koreader/contrib has the door wide open to things that don't meet the coding style or the goals of the main project btw.

**Comment by hius07** [comment]
"I've come back to say I do not contribute to KOReader"

**Comment by poire-z** [comment]
(For history, OP started with that idea 4 years ago: #6255)

**Comment by pazos** [comment]
> The plugin acts as a kind of plugin to the ReaderDictionary plugin: upon clicking on a word or name in the reader, if a xray item is found that case sensitive contains (part of) the selected text, the xray item viewer will be shown instead of the Dictionary results dialog.

Questions:

Are visual clues available on the reader to mark x-ray words when they're not clicked?

If the answer is yes, could you share a screenshot of a reader page with at least one enhanced word?.

**Comment by smartscripts-nl** [comment]
No, I've not implemented visual clues in the text. Would be tricky to realise, I think.

**Comment by Commodore64user** [comment]
Don't mind me, I'll just sit over here in the corner with some popcorn watching how this beef unfolds.

**Comment by smartscripts-nl** [comment]
It turns out to be possible to mark paragraphs with xray items in ReaderView. After visiting the same page several times, the markers for that page, though drawn, are nevertheless not shown anymore. Don't know why. But because there are so many markers in my case, I see that as a feature...
![reader-view-xray-markers](https://github.com/user-attachments/assets/d92f5d04-539c-4dc1-b4ee-fb6fe339440e)

**Comment by smartscripts-nl** [comment]
It turns out to be possible to mark paragraphs with xray items in ReaderView. After visiting the same page several times, the markers for that page, though drawn, are nevertheless not shown anymore. Don't know why. But because there are so many markers in my case, I see that as a feature...
![reader-view-xray-markers](https://github.com/user-attachments/assets/a17b3c71-854c-4558-90d3-69a2bcdf59c5)

**Comment by smartscripts-nl** [comment]
It is possible to make the first line of paragraphs marked with a xray icon tappable. By doing so the users opens a popup with xray items information in the order they were found in the paragraph. At the bottom of the popup are buttons with which the user can edit the xray items found.
![tappable-paragraphs](https://github.com/user-attachments/assets/6560f7df-0d4d-46fc-b4dc-11dd1d0e3bef)
![tappable-paragraphs-popup](https://github.com/user-attachments/assets/3144395e-e561-46c4-a64e-e60822154533)

**Comment by smartscripts-nl** [comment]
By the way: it is possible in special cases to navigate from text block with heading to text block with heading in TextViewer. 

1. The trick is to instantiate TextViewer not only with a text prop, but also with a table containing the text blocks of which that text consists.
2. The text blocks must begin with a characteristic text fragment, i.e. "header". E.g. in my example the icons at the start of the headings, followed by some heading text.
3. You can then use those blocks to derive the search text from their "headers".
4. Optionally you can set the starting char from which to start the search to just before the position of the last char of the previous text block (computed from the lengths of all previous blocks taken together). If the headings are unique enough, this step could probably be skipped.
5. Probably, instead of entire text blocks, you also could instantiate TextViewer with text and unique header texts, then to be used as a kind of index to jump from header to header.

https://github.com/user-attachments/assets/75f5759c-6aea-42e0-886a-0e15cf8429b4

**Comment by smartscripts-nl** [comment]
It _is_ possible to have a kind of toc for TextViewer, like I suggested in the previous post.

https://github.com/user-attachments/assets/ecce6598-2dd6-469e-82fe-1ac7a97df59f

**Comment by smartscripts-nl** [comment]
Another idea, to make the functionality more alike to that on Kindles: buttontable to list all items (alles), only persons (personen) or only terms (begrippen). This buttontable will only be injected if indeed there are items of both the person and the term type.

https://github.com/user-attachments/assets/a8354dff-570d-4dcc-a24a-2d1affe35eab

**Comment by smartscripts-nl** [comment]
And indicating the type of hit (full name, first name, last name, partial hit, via alias, via linked term) in the page text with icons:

![xray-items-on-page-with-icons-to-indicate-type-of-hit](https://github.com/user-attachments/assets/84703d53-8734-45d7-9d40-861fecce5d6d)
![explanation-of-hit-types](https://github.com/user-attachments/assets/0bf92c44-bdff-4243-a5f1-d8f05b433c55)

**Comment by pazos** [comment]
It trully looks like an extensive plugin, @smartscripts-nl.

It would be wonderful if you could publish the source code somewhere and let other interested people to try it as is. It doesn't mean you'll need to support the code you're releasing, of course.

I'm aware of other very extensive plugins, like https://github.com/billiam/hardcoverapp.koplugin that reach its target audience with a github repo and a post in https://www.mobileread.com/forums/forumdisplay.php?f=276

Please consider that :)

**Comment by pazos** [comment]
Thanks.

I'm going to convert this into a discussion, as it is more a "show and tell" than a FR for this repo.
